### ISSUE TITLE
AArch64 backend crash with -fsanitize=memtag-stack -mspeculative-load-hardening

### ISSUE THREAD

**Issue by ostannard** [llvm/llvm-project#61830]
This code causes a crash in the AArch64 backend when compiled with `-fsanitize=memtag-stack -mspeculative-load-hardening`:

```c
void b(int*);
void c() {
  int d[48];
  b(d);
}
```

```console
$ /work/llvm/build/bin/clang --target=aarch64--none-eabi -march=armv8.5-a+memtag -c test.c -O1 -fsanitize=memtag-stack -mspeculative-load-hardening
clang: /work/llvm/llvm-project/llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp:212: bool {anonymous}::AArch64SpeculationHardening::endsWithCondControlFlow(llvm::MachineBasicBlock&, llvm::MachineBasicBlock*&, llvm::MachineBasicBlock*&, llvm::AArch64CC::CondCode&) const: Assertion `analyzeBranchCondCode.size() == 1 && "unknown Cond array format"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /work/llvm/build/bin/clang --target=aarch64--none-eabi -march=armv8.5-a+memtag -c test.c -O1 -fsanitize=memtag-stack -mspeculative-load-hardening
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'test.c'.
4.      Running pass 'AArch64 speculation hardening pass' on function '@c'
 #0 0x0000557dde7bfb3f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/work/llvm/build/bin/clang+0x7f61b3f)
 #1 0x0000557dde7bd87c llvm::sys::CleanupOnSignal(unsigned long) (/work/llvm/build/bin/clang+0x7f5f87c)
 #2 0x0000557dde7015f8 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007fcb91e0e420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007fcb918ab00b raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007fcb9188a859 abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007fcb9188a729 get_sysdep_segment_value /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007fcb9188a729 _nl_load_domain /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007fcb9189bfd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #9 0x0000557ddbe70762 (anonymous namespace)::AArch64SpeculationHardening::instrumentControlFlow(llvm::MachineBasicBlock&, bool&) (.constprop.0) AArch64SpeculationHardening.cpp:0:0
#10 0x0000557ddbe70de7 (anonymous namespace)::AArch64SpeculationHardening::runOnMachineFunction(llvm::MachineFunction&) AArch64SpeculationHardening.cpp:0:0
#11 0x0000557dddb1e65e llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#12 0x0000557dde0d9875 llvm::FPPassManager::runOnFunction(llvm::Function&) (/work/llvm/build/bin/clang+0x787b875)
#13 0x0000557dde0d9ab9 llvm::FPPassManager::runOnModule(llvm::Module&) (/work/llvm/build/bin/clang+0x787bab9)
#14 0x0000557dde0da332 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/work/llvm/build/bin/clang+0x787c332)
#15 0x0000557ddee59a1e clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/work/llvm/build/bin/clang+0x85fba1e)
#16 0x0000557ddf5b32bc clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) CodeGenAction.cpp:0:0
#17 0x0000557de0f09129 clang::ParseAST(clang::Sema&, bool, bool) (/work/llvm/build/bin/clang+0xa6ab129)
#18 0x0000557ddf5b1bf8 clang::CodeGenAction::ExecuteAction() (/work/llvm/build/bin/clang+0x8d53bf8)
#19 0x0000557ddf487329 clang::FrontendAction::Execute() (/work/llvm/build/bin/clang+0x8c29329)
#20 0x0000557ddf4091f6 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/work/llvm/build/bin/clang+0x8bab1f6)
#21 0x0000557ddf578d87 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/work/llvm/build/bin/clang+0x8d1ad87)
#22 0x0000557ddbc7bbc6 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/work/llvm/build/bin/clang+0x541dbc6)
#23 0x0000557ddbc7725a ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#24 0x0000557ddf2708fd void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#25 0x0000557dde701ae0 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/work/llvm/build/bin/clang+0x7ea3ae0)
#26 0x0000557ddf2711bf clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#27 0x0000557ddf23525c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/work/llvm/build/bin/clang+0x89d725c)
#28 0x0000557ddf235cfd clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/work/llvm/build/bin/clang+0x89d7cfd)
#29 0x0000557ddf23d9bd clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/work/llvm/build/bin/clang+0x89df9bd)
#30 0x0000557ddbc798e0 clang_main(int, char**, llvm::ToolContext const&) (/work/llvm/build/bin/clang+0x541b8e0)
#31 0x0000557ddbc89b15 main (/work/llvm/build/bin/clang+0x542bb15)
#32 0x00007fcb9188c083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#33 0x0000557ddbc71f6e _start (/work/llvm/build/bin/clang+0x5413f6e)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 17.0.0 (git@github.com:llvm/llvm-project.git 478bd0735fc094d3af37e9791df5118a402ae7a7)
Target: aarch64-unknown-none-eabi
Thread model: posix
InstalledDir: /work/llvm/build/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/test-4bd2d1.c
clang: note: diagnostic msg: /tmp/test-4bd2d1.sh
clang: note: diagnostic msg: 

********************
```

**Comment by DavidSpickett** [comment]
@kbeyls

**Comment by kbeyls** [comment]
A reduced LLVM bitcode test case is:
```
target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
target triple = "aarch64-unknown-none-eabi"

; Function Attrs: sanitize_memtag speculative_load_hardening
define void @c() #0 {
entry:
  %d = alloca [48 x i32], align 4
  call void @b(ptr %d)
  ret void
}

declare void @b(ptr)

attributes #0 = { sanitize_memtag speculative_load_hardening "target-features"="+crc,+dotprod,+fp-armv8,+lse,+mte,+neon,+ras,+rcpc,+rdm,+v8.1a,+v8.2a,+v8.3a,+v8.4a,+v8.5a,+v8a,-fmv" }
```

```
% ~/dev/llvm/build-clang-dbg/bin/llc ./reduced.ll
Assertion failed: (analyzeBranchCondCode.size() == 1 && "unknown Cond array format"), function endsWithCondControlFlow, file AArch64SpeculationHardening.cpp, line 212.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /Users/kristof/dev/llvm/build-clang-dbg/bin/llc ./reduced.ll
1.      Running pass 'Function Pass Manager' on module './reduced.ll'.
2.      Running pass 'AArch64 speculation hardening pass' on function '@c'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  llc                      0x00000001099d4b98 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 72
1  llc                      0x00000001099d50f4 PrintStackTraceSignalHandler(void*) + 28
2  llc                      0x00000001099d2f1c llvm::sys::RunSignalHandlers() + 148
3  llc                      0x00000001099d6f28 SignalHandler(int) + 252
4  libsystem_platform.dylib 0x0000000198d802a4 _sigtramp + 56
5  libsystem_pthread.dylib  0x0000000198d51cec pthread_kill + 288
6  libsystem_c.dylib        0x0000000198c8a2c8 abort + 180
7  libsystem_c.dylib        0x0000000198c89620 err + 0
8  llc                      0x0000000105105ab0 (anonymous namespace)::AArch64SpeculationHardening::endsWithCondControlFlow(llvm::MachineBasicBlock&, llvm::MachineBasicBlock*&, llvm::MachineBasicBlock*&, llvm::AArch64CC::CondCode&) const + 484
9  llc                      0x00000001051041dc (anonymous namespace)::AArch64SpeculationHardening::instrumentControlFlow(llvm::MachineBasicBlock&, bool&) + 172
10 llc                      0x000000010510333c (anonymous namespace)::AArch64SpeculationHardening::runOnMachineFunction(llvm::MachineFunction&) + 1040
11 llc                      0x00000001080be9e0 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 792
12 llc                      0x0000000108b435c4 llvm::FPPassManager::runOnFunction(llvm::Function&) + 532
13 llc                      0x0000000108b4a5cc llvm::FPPassManager::runOnModule(llvm::Module&) + 116
14 llc                      0x0000000108b43e54 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) + 664
15 llc                      0x0000000108b439f4 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 280
16 llc                      0x0000000108b4a9c0 llvm::legacy::PassManager::run(llvm::Module&) + 36
17 llc                      0x0000000104974a94 compileModule(char**, llvm::LLVMContext&) + 4680
18 llc                      0x0000000104972f08 main + 1240
19 dyld                     0x0000000198a27e50 start + 2544
zsh: abort      ~/dev/llvm/build-clang-dbg/bin/llc ./reduced.ll
```

The assertion failure happens when processing the following basic block:
```
  $x8 = ST2GPostIndex $x8, $x8(tied-def 0), 2 :: (store (s1536) into %ir.d.tag, align 16)
  $x9 = SUBXri $x9, 32, 0
  CBNZX $x9, %bb.1
```

The root cause of the issue is that for speculative load hardening to work, conditional branches must not be implemented using CB(N)Z or TB(N)Z instructions. For more details on why, see the design document at https://github.com/llvm/llvm-project/blob/0d022e7d1ff94aaf51e8ac1cea9ce0b827c92814/llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp#L46-L61

Presumably the newly added memtag-stack sanitizer does produce CB(N)Z instructions without checking whether speculative load hardening is enabled. If so, the fix for this would be to check if speculative load hardening is enabled and not produce CB(N)Z, similar to this code: https://github.com/llvm/llvm-project/blob/b066505d8835e263e19c13f2cd718b6eb8db4470/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L8340

**Comment by kbeyls** [comment]
I've put up a patch for review that fixes this at https://reviews.llvm.org/D148508

**Comment by kbeyls** [comment]
The fix was committed in 5e612bc291347d364f1d47c37f0d34eb6474b9b5
